src/kernel/src/obj/copy.rs

```diff
@@ -24,86 +27,489 @@ fn split_range(
+    src_tree: &mut LockGuard<PageRangeTree>,
+    dest_point: PageNumber,
+    src_point: PageNumber,
+    offset: usize,
+    max: usize,
+) {
+    let src_page = src_tree.get_page(src_point, false);
+    let (dest_page, _) = dest_tree.get_or_add_page(dest_point, true, |_, _| Page::new());
+    if let Some((src_page, _)) = src_page {
+        dest_page.as_mut_slice()[offset..max].copy_from_slice(&src_page.as_slice()[offset..max]);
+    } else {
+        // TODO: could skip this on freshly created page, if we can detect that. That's just an optimization, though.
+        dest_page.as_mut_slice()[offset..max].fill(0);
+    }
+}
+
+// Zero a single, partial page.
+fn zero_single(
+    dest_tree: &mut LockGuard<PageRangeTree>,
+    dest_point: PageNumber,
+    offset: usize,
+    max: usize,
+) {
+    // if there's no page here, our work is done
+    if let Some((dest_page, _)) = dest_tree.get_page(dest_point, true) {
+        dest_page.as_mut_slice()[offset..max].fill(0);
+    }
+}
+/// Copy page ranges from one object to another, preferring to share page vectors if possible.
+///
+/// In the case that a full page needs to be copied, it will likely be shared and set
+/// to copy on write. In the case that a page needs to be partially copied, we'll do a manual copy
+/// for that page. This only happens at the start and end of the copy region.
+///
+/// We allow non-page-aligned offsets, and that misalignment may differ between source and dest objects,
+/// but the kernel may have to resort to a bytewise copy of the object pages if the offsets aren't both
+/// misaligned by the same amount (e.g., if page size is 0x1000, then (dest off, src off) of (0x1000, 0x4000),
+/// (0x1100, 0x3100) will still enable COW style copying, but (0x1100, 0x1200) will require manual copy).
+///
+/// We lock the page trees for each object (in a canonical order) and ensure that the regions are
+/// remapped appropriately for any mapping of the objects. This ensures that the source object is
+/// "checkpointed" before copying, and that the destination object cannot be read in the region being
+/// overwritten until the copy is done.
 pub fn copy_ranges(
     src: &ObjectRef,
-    src_start: PageNumber,
+    src_off: usize,
     dest: &ObjectRef,
-    dest_start: PageNumber,
-    length: usize,
+    dest_off: usize,
+    byte_length: usize,
 ) {
-    let (src_tree, mut dest_tree) = crate::utils::lock_two(&src.range_tree, &dest.range_tree);
+    let src_start = PageNumber::from_offset(src_off);
+    let dest_start = PageNumber::from_offset(dest_off);
+
+    let start_offset = src_off % PageNumber::PAGE_SIZE;
+    let end_offset = (src_off + byte_length) % PageNumber::PAGE_SIZE;
+
+    // Number of pages that will be touched, including partial pages.
+    // By subtracting the partial pages, we are left with the full pages,
+    // and then we can add in how many partial pages we'll be copying.
+    let nr_pages: usize = byte_length.saturating_sub(start_offset + end_offset)
```

Comment:
How does the sub of `start_offset` from `byte_length` make sense? Would it not be the remaining bytes we care about (`PAGE_SIZE - start_offset`)?

Reply:
Yes you're right, nice catch. Will update.